scopesim/source/source_utils.py

```diff
@@ -1,163 +1,219 @@
+            elif all(isinstance(spec, np.ndarray) for spec in spectra):
+                yield from _from_arrays(spectra)
+            else:
+                raise ValueError(
+                    "If given as an iterable, spectra must consist of all "
+                    "synphot spectra or all arrays")
+            return
+
+        if isinstance(spectra, np.ndarray):
+            if spectra.ndim == 1:
+                yield _synphotify(spectra)
+            elif spectra.ndim == 2:
+                yield from _from_arrays(spectra)
+            else:
+                raise ValueError(
+                    "If given as an array, spectra must have either 1 (single "
+                    "flux list) or 2 (flux of multiple spectra) dimensions, "
+                    f"but {spectra.ndim} were found.")
+            return
+
+    return list(_get_list())
+
+
+# FIXME: typing: This should work with the more general (and true, since
+#        conversion is done anyway) Quantity["length"] (or "area" resp.), but
+#        doing so currently causes a NameError. Not sure what's going on.
+def photons_in_range(
+        spectra: SourceSpectrum,
+        wave_min: u.Quantity[u.um] | float,
+        wave_max: u.Quantity[u.um] | float,
+        area: Optional[Union[u.Quantity[u.m**2], float]] = None,
+        bandpass: Optional[SpectralElement] = None,
+) -> Union[u.Quantity[u.ph * u.s**-1 * u.m**-2], u.Quantity[u.ph * u.s**-1]]:
     """
+    Integrate photons from spectrum in given wavelength range.
 
     Parameters
     ----------
-    spectra
-    wave_min
-        [um]
-    wave_max
-        [um]
-    area : Quantity
-        [m2]
-    bandpass : SpectralElement
-
+    spectra : SourceSpectrum
+        Input spectrum.
+    wave_min : Union[u.Quantity["length"], float]
+        Minimum wavelength. If float, assumes um.
+    wave_max : Union[u.Quantity["length"], float]
+        Maximum wavelength. If float, assumes um.
+    area : Optional[Union[u.Quantity["area"], float]], optional
+        Area to multiply with. If float, assumes m**2. The default is None.
+    bandpass : Optional[SpectralElement], optional
+        Filter to take into account, if any. The default is None.
 
     Returns
     -------
-    counts : u.Quantity array
+    counts : astropy.units.Quantity
+        Either in ph/s/m**2 or just ph/s (if area was given).
 
     """
-
-    if isinstance(wave_min, u.Quantity):
-        wave_min = wave_min.to(u.Angstrom).value
-    else:
-        wave_min *= 1E4
-
-    if isinstance(wave_max, u.Quantity):
-        wave_max = wave_max.to(u.Angstrom).value
-    else:
-        wave_max *= 1E4
+    # Note: Assuming um if given as float.
+    wave_min = (wave_min << u.um << u.Angstrom).value
+    wave_max = (wave_max << u.um << u.Angstrom).value
+    # Note: There appear to be some float shenanigans going on here, but
+    # rounding produces an error in the spectrum evaluation. Not sure what's
+    # going on, maybe it's fine as-is.
```

Comment:
This is probably NOT fine as it is. The problem is:
```
>>> (1 << u.um << u.Angstrom).value
9999.999999999998
```
The `1E4` conversion was ugly, but correct.

It might be that in this particular case everything is fine, but I think we should not rely on that.

This incorrect conversion also causes all kinds of problems with synphot.

Reply:
I've reconsidered. We should make this change.

This unit conversion could indeed go wrong. However, it already would go wrong if fractional wavelengths are used, e.g. `wavemin=1333.3`. Our code should not depend on float values being integer.

This means that even if we would hardcode this 1E4 here (and similarly elsewhere), that the code will still be broken in general. So we should not use such hacks and instead just let astropy handle it, as you did.

Per extension, it should not cause any problems if a value of 1000.0 becomes 999.9999999999 or 1000.000000001:
- Any (synphot) code that computes overlaps should be resilient to such rounding errors and just ignore it.
- Any plotting/printing code should just print 1000.0. (At least for UX, not for debugging.)

Perhaps we should find some floating point numbers that you cannot faithfully multiply/divide by (e.g.) 1E4. Maybe one number that rounds up and one that rounds down. Then we should use those in the basic_instrument test package to detect any errors that such a conversion causes. Or maybe have some test irdb packages that uses weird units like wavelengths in parsec and such.